### PR TITLE
Add agency proofs to the `Protocol` instance

### PR DESCRIPTION
**Pull request by coot**
TODO: the docs needs to be checked.

### CODE REVIEW

File: typed-protocols/src/Network/TypedProtocol/Core.hs [hunk 1]
```diff
@@ -166,6 +170,70 @@ module Network.TypedProtocol.Core (
...
+-- For this protocol there is only one state in which the client has agency,
+-- the idle state. By pattern matching on the state token for the server
+-- agency we can list all the cases in which the server also has agency for
+-- the idle state. There are of course none of these so we give the empty
+-- set of patterns. GHC can check that we are indeed correct about this.
+-- This also requires the @EmptyCase@ language extension.
+--
+-- To get this completeness checking it is important to compile modules
+-- containing these lemmas with @-Wincomplete-patterns@, which is implied by
+-- @-Wall@.
```
Comment (avieth) [inline typed-protocols/src/Network/TypedProtocol/Core.hs]: Good explanation! Should also put in a derived lemma which switches the client and server parts.

```Haskell
proofByContradiction_ServerExclusive :: ServerHasAgency st -> ClientHasAgency st -> Void
proofByContradiction_ServerExclusive = \server -> \client -> 
  proofByContradiction_ClientAndServerHaveAgency client server
```

File: typed-protocols/src/Network/TypedProtocol/Core.hs [hunk 2]
```diff
@@ -209,6 +277,32 @@ class Protocol ps where
...
+  proofByContradiction_NobodyAndClientHaveAgency
+    :: forall (st :: ps).
+      NobodyHasAgency st
+    -> ClientHasAgency st
+    -> Void
+
+  -- | Lemma that if the nobody has agency for a state, there are no
+  -- cases in which the server has agency for the same state.
+  --
+  proofByContradiction_NobodyAndServerHaveAgency
```
Comment (avieth) [inline typed-protocols/src/Network/TypedProtocol/Core.hs]: I think the names for these 3 should be slightly different. `ClientAndServerHaveAgency` kind of suggests that they both have agency. How about `proofByContradiction_ClientServerExclusive`?

Comment (dcoutts) [reply typed-protocols/src/Network/TypedProtocol/Core.hs]: Right, I was trying to find a short way of saying "nobody and server have agency leads to a contradition". I'll merge and do some renaming after.

File: typed-protocols/src/Network/TypedProtocol/Core.hs (hunk 1 above)
Comment (dcoutts) [reply typed-protocols/src/Network/TypedProtocol/Core.hs]: Thanks, that's what I added previously, based on your critique.

Not a bad idea, adding a flipped one. Currently we just use it flipped in the `connect` proofs.

**Pull request merged**